MDEV-10654 add: Support IN, OUT, INOUT qualifiers for stored functions

ManoharKB opened the pull request:
Implemented IN, OUT, INOUT qualifiers for stored function parameters and added input and ouput parameter binding for OUT/INOUT parameters.

<!--
Thank you for contributing to the MariaDB Server repository!

You can help us review your changes faster by filling this template <3

If you have any questions related to MariaDB or you just want to
hang out and meet other community members, please join us on
https://mariadb.zulipchat.com/ .
-->

<!--
If you've already identified a https://jira.mariadb.org/ issue
that seems to track this bug/feature, please add its number below.
-->
- [x] *The Jira issue number for this PR is: MDEV-10654*

<!--
An amazing description should answer some questions like:
-->
## Description
1. Current MariaDB 10.7 does not support IN, OUT, INOUT parameter qualifiers in CREATE FUNCTION, for Oracle compatibility (When running with sql_mode=ORACLE)
2. This patch will allows to use IN, OUT, INOUT parameter qualifiers in CREATE FUNCTION, pass input values through IN, INOUT parameters and return by reference through OUT, INOUT parameters during execution. This is supported only with sql_mode=ORACLE.
3. To maintain compatibility with Oracle, Stored Function with OUT/INOUT parameter cannot be called directly within SQL query (SELECT, INSERT, UPDATE & DELETE).
4. sql/sql_yacc.yy : Supporting IN/OUT/INOUT qualifiers  in CREATE FUNCTION
5. sql/sp_head.cc : Input and output parameter binding and supporting OUT/INOUT parameters (return by reference).
6. sql/share/errmsg-utf8.txt : Added error message to restrict calling function with OUT/INOUT qualifiers from SQL query.

## How to call function

Ex1:
`SET sql_mode='ORACLE';
DELIMITER $$
CREATE OR REPLACE PACKAGE pkg2
AS
  FUNCTION add_func3 (a IN INT, b IN INT, c OUT INT) RETURN INT;
END;
$$
DELIMITER ;

set @a = 2;
set @b = 3;
set @c = 0;
select pkg2.add_func3(@a, @b, @c);
`

Ex2:
`
SET sql_mode='ORACLE';
DELIMITER $$
CREATE OR REPLACE PACKAGE pkg2
AS
  FUNCTION func(a OUT INT) RETURN INT;
END;
$$
DELIMITER ;

set @a = 0;
INSERT INTO Persons SELECT 5, 'EEE', PKG2.func(@a);
`
## How can this PR be tested?
This patch has been tested with below scenarios:

- CREATE FUNCTION with IN, OUT, INOUT parameter qualifiers and without parameter qualifiers.
- Function (no qualifiers, IN, OUT & INOUT) calls from SQL query (SELECT, INSERT, UPDATE  & DELETE).
- Function (no qualifiers, IN, OUT & INOUT) called from stored procedure.
- Function called from stored procedure with different order of parameters (between IN, OUT, INOUT) between procedure and function.
- Call procedure inside function (without qualifier, IN, OUT & INOUT).
- SQL statements (SELECT, INSERT, UPDATE  & DELETE) within function (no qualifiers, IN, OUT & INOUT)
- Procedure ---calls---> function ---calls---> SQL statement
- Trigger ---calls---> function ---calls---> SQL statement
- Trigger ---calls---> procedure ---calls---> function ---calls---> SQL statement

In the above scenarios, functions with and without parameter qualifiers are tested. For SQL statements SELECT, INSERT, UPDATE  & DELETE are tested.

<!--
Tick one of the following boxes [x] to help us understand
if the base branch for the PR is correct
-->
## Basing the PR against the correct MariaDB version
- [x] *This is a new feature and the PR is based against the latest MariaDB development branch*
- [ ] *This is a bug fix and the PR is based against the earliest branch in which the bug can be reproduced*

<!--
You might consider answering some questions like:
1. Does this affect the on-disk format used by MariaDB?
2. Does this change any behavior experienced by a user
   who upgrades from a version prior to this patch?
3. Would a user be able to start MariaDB on a datadir
   created prior to your fix?
-->
## Backward compatibility
Backward compatibility for stored function is maintained. That is, function can be created without specifying parameter qualifiers. So, there is no effect on the existing behavior.

grooverdan commented:
Thanks @ManoharKB. If you could include some test cases described in https://lists.launchpad.net/maria-developers/msg12972.html that would be appreciated.

grooverdan commented:
The other bit I notice is (maybe) missing is `show create function f` to include this extra sql syntax in its output

abarkov commented:
Hello. Thanks for your contribution. It generally looks good. However I have some suggestions to improve the code.

1. I tried this script:

```
SET sql_mode='ORACLE';
DELIMITER $$
CREATE OR REPLACE PACKAGE pkg2 AS
  FUNCTION add_func3 (a IN INT, b IN INT, c OUT INT) RETURN INT;
END;
$$
DELIMITER ;

DELIMITER $$
CREATE OR REPLACE PACKAGE BODY pkg2
AS
  FUNCTION add_func3 (a IN INT, b IN INT, c OUT INT) RETURN INT AS
  BEGIN
    c:= a+b;
    RETURN 1;
  END;
END;
$$
DELIMITER ;

set @A = 2;
set @b = 3;
set @c = 0;
select pkg2.add_func3(@A, @b, @c);
```
It returns this error:
```
ERROR 4186 (HY000): OUT or INOUT argument 3 for function pkg2.add_func3 is not allowed here
```
I think that a simple SELECT without tables should be allowed.

2. We'd like OUT parameters to be available in both PL/SQL and SQL/PSM dialects.
3. Please add tests into mysql-test/main and into suite/compat/oracle/t
4. We usually try to avoid copying the code. Please add new sp_head methods and reuse it in both sp_head::execute_procedure() and sp_head::execute_function(), instead of copying

ManoharKB commented:
@grooverdan @abarkov Thank you for the review.
I would like to clarify 1st point @abarkov mentioned.
> I think that a simple SELECT without tables should be allowed.

According to the below Oracle link, it is not allowed.
https://docs.oracle.com/cd/B19306_01/server.102/b14200/statements_5009.htm
Under section "Restrictions on User-Defined Functions "
I have also unit tested this case in Oracle - it throws "Error: ORA-06572: Function … has out arguments"

And I will look into other review points from @abarkov and @grooverdan 

Thank you.

abarkov commented:
In the "How to call function" examples you use these statements:

```
select pkg2.add_func3(@A, @b, @c);
INSERT INTO Persons SELECT 5, 'EEE', PKG2.func(@A);
```
But these two statements do not work - they return the error.

ManoharKB commented:
> In the "How to call function" examples you use these statements:
> 
> ```
> select pkg2.add_func3(@A, @b, @c);
> INSERT INTO Persons SELECT 5, 'EEE', PKG2.func(@A);
> ```
> 
> But these two statements do not work - they return the error.

Sorry for the confusion. In this section, I meant only to explain how to call function, but unintentionally took error returning examples.
Both the above statements return error, which is correct.
I will upload my unit test cases (Oracle vs MariaDB) probably today.

ManoharKB commented:
I have added test cases to and pushed the changes.
1. mysql-test/suite/compat/oracle/t/func_inout.test
2. mysql-test/suite/compat/oracle/r/func_inout.result
Commit ID: 771108c9261e4b2a6f26af1d58550789a9e447c7

As the ticket description says, this fix only supports when sql_mode='ORACLE';

Thank you.

ManoharKB commented:
@abarkov 
> We usually try to avoid copying the code. Please add new sp_head methods and reuse it in both sp_head::execute_procedure() and sp_head::execute_function(), instead of copying

Refactored the source code in commit 3c36979b20d7403893fe10ff08e8e9da0fd51cd2 and tested.
Please review.
Thank you.

ManoharKB commented:
@abarkov Thank you for the review and feedback.

1. Fixed coding style issues
2. Renamed func_inout.test to sp-inout.test (similarly for .result)
3. Squashed all commits.

Please review.
Thank you.

abarkov commented:
[addon.diff.txt](https://github.com/MariaDB/server/files/7537878/addon.diff.txt)

abarkov commented:
Please apply the patch addon.diff.txt I attached to this PR.
Changes:
1. The nctx parameter to bind_input_params() is now `sp_rcontext*` instead of `sp_rcontext**`
2. Fixed the wrong use of DBUG_ENTER/DBUG_RETURN. It did not compile with Debug build mode.
3. Added a test checking the use of the package body internal variables as OUT parameters

abarkov commented:
I have questions about these two fragments:

1.  In sp_head::execute_function()
```
  /* Pass arguments. */
  for (arg_no= 0; arg_no < argcount; arg_no++)
  {
    /* Arguments must be fixed in Item_func_sp::fix_fields */
    DBUG_ASSERT(argp[arg_no]->fixed());

    if (!argp[arg_no]) // QQ: why is this needed?
      break;
```
I think the old code have this test. Why is this needed?
Note, if it's really needed, it's on the wrong place anyway, shouldn't it be before the DBUG_ASSERT()?



2. sp_head::bind_output_params()
```
   sp_rcontext *tmp_ctx= octx;
  if (is_function && !octx) // QQ: why this needed?
  {
    tmp_ctx= nctx;
```
Why do we need this tmp_ctx variable?
If octx is NULL, it means the function is called directly (not from another stored routine), e.g. like this:
```
mysql> SET @var=func_with_out_param(@b);
```
But this also mean that the OUT argument is not an stored routine variable, so the sp_rcontext pointer passed to `srp->set_value()`  can probably be NULL.

In the above example srp points to an `Item_func_get_user_var` instance. So Item_func_get_user_var::set_value() is called. And it can perfectly accept NULL as the `sp_rcontext*` - it does not even use it.

I commented this line out:
```
 //   tmp_ctx= nctx;
```
and `mtr sp-inout` still worked without problems.

Do you have an example where we really need to replace octx==NULL to some not-NULL value?

abarkov commented:
Oh, by the way, can you please rename:
- `bind_input_params()` to `bind_input_param()`
- `bind_output_params()` to `bind_output_param()`

These two methods handle one parameter at a time, so the plural noun sounds confusing.
Thanks.

ManoharKB commented:
@abarkov Thank you for the review and patch.

I have committed the below changes to the branch. Commit ID: a0a3138efd83e17e03066d26b58bb60ba8cb87ae
1. Applied your patch addon.diff.txt
2. >> if (!argp[arg_no]) // QQ: why is this needed?
    I had added it to be consistent with execute_procedure(), but this is not needed specifically, DBUG_ASSERT() before that will suffice. I removed this line in the current commit.
3. >>  sp_rcontext *tmp_ctx= octx;
    >> if (is_function && !octx) // QQ: why this needed?
    During initial debugging with gdb, I encountered error when octx was NULL (may be during direct call) and may be my understanding was wrong wrt octx then. I re-executed all scenarios I know, but could not find any issue without tmp_ctx. So, I have removed this line in the current commit.
4. Renamed bind_input_params() and bind_output_params()
5. Squashed all commits.

Ran 'mtr sp-input' without issues.

Thank you.

abarkov commented:
Hello ManoharKB,

I have merged your patch to this MariaDB Server branch:

https://github.com/MariaDB/server/tree/preview-10.8-MDEV-10654-inout

After testing it by our Quality Assurance team the patch will be merged to the main 10.8 branch.

Thank you very much for your great contribution, and for addressing review suggestions!


This PR will stay open for the time being. We'll close it when the patch gets merged to 10.8.

grooverdan commented:
all merged into 10.8 branch now - 4572dc23f7d83dfffc9cf2577cea96a034c5332c

Thanks for your contribution!
